server/src/main/java/org/opensearch/index/translog/transfer/BlobStoreTransferService.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.index.translog.transfer;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.opensearch.action.ActionListener;
+import org.opensearch.action.ActionRunnable;
+import org.opensearch.common.blobstore.BlobPath;
+import org.opensearch.common.blobstore.BlobStore;
+import org.opensearch.index.translog.transfer.FileSnapshot.TransferFileSnapshot;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.concurrent.ExecutorService;
+
+/**
+ * Service that handles remote transfer of translog and checkpoint files
+ *
+ * @opensearch.internal
+ */
+public class BlobStoreTransferService implements TransferService {
+
+    private final BlobStore blobStore;
+    private final ExecutorService executorService;
+
+    private static final Logger logger = LogManager.getLogger(BlobStoreTransferService.class);
+
+    public BlobStoreTransferService(BlobStore blobStore, ExecutorService executorService) {
+        this.blobStore = blobStore;
+        this.executorService = executorService;
+    }
+
+    @Override
+    public void uploadBlobAsync(
+        final TransferFileSnapshot fileSnapshot,
+        Iterable<String> remoteTransferPath,
+        ActionListener<TransferFileSnapshot> listener
+    ) {
+        assert remoteTransferPath instanceof BlobPath;
```

Comment:
Why not make the parameter type a BlobPath to let the compiler do this check for you?

Reply:
Since the interface is generic and not tied to a store, could be extended to a streaming store as well

Reply:
Since you currently only have one instance of a TransferService, I would recommend not introducing an interface and do things like this. It's easy enough to refactor out an interface if/when you need it, but until you really need it you can only make educated guesses about what the generic interface should look like. Not a huge deal, I'll defer to your preference.